### ISSUE TITLE
make tuning frequency customizable in midi_to_hz() function

### ISSUE THREAD

**Issue by nullmightybofo** [librosa/librosa#977]
#### Description

Make `midi_to_hz()` accept a custom tuning frequency. Currently, 440.0 is hard coded.

I think it would be nice to have `midi_to_hz()` accept a different tuning frequency. That way, it would be easy to generate frequencies for notes following that tuning, etc.

I.e., turn `midi_to_hz(notes)` to `midi_to_hz(notes, A4=440.0)`.

And change: 

 return 440.0 * (2.0 ** ((np.asanyarray(notes) - 69.0)/12.0))

to: 

 return A4 * (2.0 ** ((np.asanyarray(notes) - 69.0)/12.0)).

(I am not sure whether librosa has "agreed upon" a note naming convention for tuning. Is C4 the middle C (261 Hz in equal temperament) in librosa? I think so.)

Code here:
https://librosa.github.io/librosa/_modules/librosa/core/time_frequency.html#midi_to_hz

The change would add the the functionality of choosing tuning frequency to other functions that depend on `midi_to_hz()` as well, such as `note_to_hz()`:
https://librosa.github.io/librosa/_modules/librosa/core/time_frequency.html#note_to_hz

I can write a PR if this seems helpful.

Thanks and BR!

**Comment by lostanlen** [comment]
Yes, this would be a nice to have.
In the wake of #930 (which closed #925), i would be wary of the keyword `A4=440` though. Experience has shown that it causes some API headaches. 
For consistency across the library, perhaps it would be good to define it in terms of a `tuning` keyword, measured in cents with respect to the A440 tuning standard. This would resolve your question as to whether it's better to express tuning in terms of A4, middle C, or other, etc.

I hope this helps!

**Comment by nullmightybofo** [comment]
Thanks for getting back so quick!

A couple of thoughts from my side:

- After briefly looking at #925 and #930, it seems all this is more involved, and it seems good to have that standardized across the whole library.

- I think though, as an interface to a user, allowing for setting the tuning frequency for A4 in terms of Hz is very useful, as is setting via a cent value (1/100 of semitone in equal temp.) in reference to A4=440. Using fractions of a bin is imo the least intuitive; maybe it makes sense to use this internally, for exchange between functions, but it is not really easy to grasp, I'd say. All these can be converted to each other anyway, right? (In the simple functions like `hz_to_note`, etc., of course, bins are not involved at all, so only frequencies and cents are applicable.)

- Maybe there should be a version of [estimate_tuning](https://librosa.github.io/librosa/generated/librosa.core.estimate_tuning.html?highlight=estimate_tuning#librosa.core.estimate_tuning) that returns a more human-readable unit for tuning?

- Should there then be a standardized unit for tuning in librosa, and some conversion functions that allow a tuning frequency in Hz as input? Like so: `note_to_hz(69, tuning=tuning_from_hz(443))` ?

**Comment by bmcfee** [comment]
Circling back on this, I'd like to include the ref-pitch to tuning conversion utility in the next major revision.  I think this is pretty straightforward as an API, and for completeness, it's good to include the reverse conversion as well.

```python
tuning = refA_to_tuning(A=440, bins_per_octave=12)
ref_A = tuning_to_refA(tuning=0, bins_per_octave=12)
```

The BPO needs to be included in the API to allow consistent interpretation of tuning as fractions of a logarithmic frequency bin when octaves are subdivided into more or fewer bins.  The default of 12 results in tuning in cents (as expected), but this gets really inconvenient when mixed with, eg, 24- or 36-BPO bases used in CQTs.  (Not to mention other non-standard systems like 19tet or whatever.)

The math isn't terribly complicated, and already exists (implicitly) in `pitch_tuning` and `hz_to_octs` / `octs_to_hz`:
```python
tuning = bins_per_octave * (np.log2(ref_A) - np.log2(440.0))
ref_A = 440.0 * 2.0**(tuning / bins_per_octave)
```

I think this would be a good beginner issue, if someone wants to take it on.

**Comment by jlw365** [comment]
Hi @bmcfee, would be happy to take this issue on. Just wanted to confirm a few things based on the above discussion:

- For this initial PR, based on your recent comment, sounds like you'd like to start by adding the two ref-pitch to tuning conversion util methods described above. Do you want to wait on adding a `units` parameter of sorts in additional methods as discussed [here](https://github.com/librosa/librosa/issues/977#issuecomment-525722692) then?
- Confirming that these would be added to the core [time_frequency module](https://librosa.github.io/librosa/_modules/librosa/core/time_frequency.html) and not general [utils](https://librosa.github.io/librosa/util.html)?

Thanks!

**Comment by bmcfee** [comment]
Thanks @jlw365 !



>     * Do you want to wait on adding a `units` parameter of sorts in additional methods as discussed [here](https://github.com/librosa/librosa/issues/977#issuecomment-525722692) then?

Yeah, let's hold off on that one, since it might require a good amount of integration and auditing to make sure it's consistently implemented across the package.  I think for now, manually converting ref to tuning is totally fine.  Anyone that wants to use ref pitches instead can afford the extra line of code to convert. :grin: 



>     * Confirming that these would be added to the core [time_frequency module](https://librosa.github.io/librosa/_modules/librosa/core/time_frequency.html), with unit tests in the respective location, and not in general [utils](https://librosa.github.io/librosa/util.html)?

Yup, these would fit in with all the other unit conversions.



>     * Thoughts on naming them `ref_pitch_to_tuning` and `tuning_to_ref_pitch`?

How about `A4_to_tuning` (and vice versa)?  `ref_pitch` is a bit wordy, and might not be precise enough for readers to directly grok from skimming the documentation. `A440_to_tuning` would be good because A440 is more descriptive than "reference pitch", but bad because if the input was actually 440, we wouldn't be calling the function at all. `A4_to_tuning` allows for any frequency corresponding to A4, but I think most readers looking for this kind of function would immediately understand what it is.

I'm open to other suggestions though!  Maybe we can go with this, and punt further discussion to code review?
